File: services/runlog/sla.py
"""SLA y regla dura de timeout (§7.3).

    Un HITL vencido escala o expira. Nunca auto-aprueba.

Toda la aritmetica es de horas habiles, porque un SLA de 4 horas que corre un sabado a las
once de la noche no es un SLA: es una trampa.

DE DONDE SALEN LOS NUMEROS. De ningun lugar de este archivo. Hasta v3.0.1 la jornada y los
plazos estaban aqui en duro, y `authority-gate.yaml` declaraba los mismos plazos sin que
nadie los leyera: cambiar el SLA en la politica no cambiaba nada. Ahora hay una sola fuente
por cosa, y este modulo la consume:

    registry/policies/calendario-laboral.yaml   huso, jornada, dias habiles, festivos
    registry/policies/authority-gate.yaml       hitl.sla: plazo y consecuencia por criticidad

Cambiar un horario o un plazo es editar un YAML y abrir un PR. No se toca este archivo.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import date, datetime, time, timedelta, timezone
from functools import lru_cache
from pathlib import Path
from typing import Literal

import yaml

from services.common.errors import ErrorDeValidacion

RAIZ = Path(__file__).resolve().parent.parent.parent
CALENDARIO = RAIZ / "registry" / "policies" / "calendario-laboral.yaml"
GATE = RAIZ / "registry" / "policies" / "authority-gate.yaml"

ACCIONES = ("escalar", "expirar", "bloquear")


class NuncaAutoAprueba(AssertionError):
    """Se intento resolver un vencimiento aprobando. Es la unica salida prohibida.

    Se levanta al CARGAR la politica, no al aplicarla. Un `al_vencer: aprobar` escrito en
    el YAML tiene que reventar el arranque, no descubrirse el dia que un HITL vence.
    """


# --- la politica, cargada -------------------------------------------------


@dataclass(frozen=True)
class Calendario:
    """El horario de oficina declarado en calendario-laboral.yaml."""

    huso: timezone
    apertura: time
    cierre: time
    dias_habiles: tuple[int, ...]
    festivos: frozenset[date]
    calibrado: str = "parcial"
    version: str = "v1"

    @property
    def horas_por_dia(self) -> float:
        """Un dia habil dura lo que dure la jornada. No se supone: se resta."""
        abre = self.apertura.hour + self.apertura.minute / 60
        cierra = self.cierre.hour + self.cierre.minute / 60
        return cierra - abre


def _hora(texto: object, campo: str) -> time:
    try:
        return time.fromisoformat(str(texto))
    except ValueError as exc:
        raise ErrorDeValidacion(f"{campo} no es una hora valida: {texto!r}", campo=campo) from exc


@lru_cache(maxsize=8)
def cargar_calendario(ruta: str | None = None) -> Calendario:
    destino = Path(ruta) if ruta else CALENDARIO
    if not destino.is_file():
        raise ErrorDeValidacion(f"no existe el calendario laboral: {destino}", campo="calendario")
    datos = yaml.safe_load(destino.read_text(encoding="utf-8")) or {}

    huso_cfg = datos.get("huso") or {}
    jornada = datos.get("jornada") or {}
    festivos_cfg = datos.get("festivos") or {}

    apertura = _hora(jornada.get("apertura", "09:00"), "apertura")
    cierre = _hora(jornada.get("cierre", "18:00"), "cierre")
    if apertura >= cierre:
        raise ErrorDeValidacion(
            f"la jornada abre a las {apertura} y cierra a las {cierre}: nunca avanzaria el reloj",
            campo="jornada",
        )

    # `or` no sirve aqui: una lista vacia es falsy y quedaria convertida en el valor
    # por defecto. "No lo declare" y "declare que no hay ninguno" no son lo mismo.
    declarados = jornada.get("dias_habiles")
    dias = tuple(int(d) for d in (declarados if declarados is not None else [0, 1, 2, 3, 4]))
    if not dias:
        raise ErrorDeValidacion("no hay ni un dia habil: ningun SLA podria vencer", campo="dias_habiles")
    if any(d < 0 or d > 6 for d in dias):
        raise ErrorDeValidacion(f"dia habil fuera de rango 0-6: {dias}", campo="dias_habiles")

    # PyYAML ya convierte `2026-09-16` en un date. Un texto tambien se acepta, porque el
    # dia que alguien pegue la lista entre comillas no tiene por que fallar en silencio.
    festivos = frozenset(
        f if isinstance(f, date) else date.fromisoformat(str(f))
        for f in (festivos_cfg.get("fechas") or [])
    )

    offset = float(huso_cfg.get("offset_horas", -6))
    return Calendario(
        huso=timezone(timedelta(hours=offset), str(huso_cfg.get("nombre") or "local")),
        apertura=apertura,
        cierre=cierre,
        dias_habiles=dias,
        festivos=festivos,
        calibrado=str(datos.get("calibrado") or "parcial"),
        version=str(datos.get("version") or "v1"),
    )


@dataclass(frozen=True)
class ReglaSLA:
    criticidad: str
    al_vencer: str
    luego: str
    horas_habiles: float | None = None
    dias_habiles: int | None = None
    resumen: str = ""

    def horas(self, calendario: Calendario) -> float:
        if self.horas_habiles is not None:
            return self.horas_habiles
        return (self.dias_habiles or 0) * calendario.horas_por_dia


@lru_cache(maxsize=8)
def cargar_sla(ruta: str | None = None) -> dict[str, ReglaSLA]:
    """Los plazos de §7.3, leidos de authority-gate.yaml -> hitl.sla."""
    destino = Path(ruta) if ruta else GATE
    if not destino.is_file():
        raise ErrorDeValidacion(f"no existe el gate de autoridad: {destino}", campo="gate")
    datos = yaml.safe_load(destino.read_text(encoding="utf-8")) or {}
    crudo = ((datos.get("hitl") or {}).get("sla")) or {}
    if not crudo:
        raise ErrorDeValidacion("authority-gate.yaml no declara hitl.sla", campo="sla")

    reglas: dict[str, ReglaSLA] = {}
    for criticidad, regla in crudo.items():
        al_vencer = str(regla.get("al_vencer") or "")
        luego = str(regla.get("luego") or al_vencer)

        # La regla dura, impuesta donde se escribe la politica y no donde se aplica.
        for accion in (al_vencer, luego):
            if accion not in ACCIONES:
                raise NuncaAutoAprueba(
                    f"hitl.sla.{criticidad} declara {accion!r}, que no es una consecuencia "
                    f"permitida. Un HITL vencido escala o expira; nunca auto-aprueba (§7.3). "
                    f"Validas: {', '.join(ACCIONES)}"
                )

        if regla.get("horas_habiles") is None and regla.get("dias_habiles") is None:
            raise ErrorDeValidacion(
                f"hitl.sla.{criticidad} no declara plazo: falta horas_habiles o dias_habiles",
                campo="sla",
            )

        reglas[str(criticidad)] = ReglaSLA(
            criticidad=str(criticidad),
            al_vencer=al_vencer,
            luego=luego,
            horas_habiles=regla.get("horas_habiles"),
            dias_habiles=regla.get("dias_habiles"),
            resumen=str(regla.get("resumen") or ""),
        )
    return reglas


# --- constantes de compatibilidad ------------------------------------------
# El resto del sistema y las pruebas importan estos nombres. Siguen existiendo, pero ya
# no son la declaracion: son el resultado de leer la politica.

_CAL = cargar_calendario()

HUSO = _CAL.huso
APERTURA = _CAL.apertura
CIERRE = _CAL.cierre
DIAS_HABILES = _CAL.dias_habiles
HORAS_POR_DIA = _CAL.horas_por_dia

#: §7.3 tal como queda tras leer authority-gate.yaml. Se conserva la forma de diccionario
#: que tenia cuando estaba en duro, para no romper a quien ya la importaba.
SLA = {
    criticidad: {
        **({"horas_habiles": r.horas_habiles} if r.horas_habiles is not None else {}),
        **({"dias_habiles": r.dias_habiles} if r.dias_habiles is not None else {}),
        "al_vencer": r.al_vencer,
        "luego": r.luego,
    }
    for criticidad, r in cargar_sla().items()
}


# --- aritmetica del calendario ---------------------------------------------


def a_local(momento: datetime, calendario: Calendario | None = None) -> datetime:
    """Lleva cualquier momento a la hora de oficina. Un naive se asume ya local."""
    cal = calendario or _CAL
    if momento.tzinfo is None:
        return momento.replace(tzinfo=cal.huso)
    return momento.astimezone(cal.huso)


def _es_dia_habil(momento: datetime, cal: Calendario) -> bool:
    """Lunes a viernes y que no sea festivo. Un festivo cuenta igual que un sabado."""
    return momento.weekday() in cal.dias_habiles and momento.date() not in cal.festivos


def es_habil(momento: datetime, calendario: Calendario | None = None) -> bool:
    cal = calendario or _CAL
    local = a_local(momento, cal)
    return _es_dia_habil(local, cal) and cal.apertura <= local.time() < cal.cierre


def _siguiente_apertura(momento: datetime, cal: Calendario) -> datetime:
    candidato = momento
    if candidato.time() >= cal.cierre:
        candidato = (candidato + timedelta(days=1)).replace(
            hour=cal.apertura.hour, minute=cal.apertura.minute, second=0, microsecond=0
        )
    elif candidato.time() < cal.apertura:
        candidato = candidato.replace(
            hour=cal.apertura.hour, minute=cal.apertura.minute, second=0, microsecond=0
        )
    # Un instante dentro de la jornada pero en sabado —o en festivo— no entra en ninguna
    # rama de arriba: lo mueve este bucle.
    while not _es_dia_habil(candidato, cal):
        candidato = (candidato + timedelta(days=1)).replace(
            hour=cal.apertura.hour, minute=cal.apertura.minute, second=0, microsecond=0
        )
    return candidato


def sumar_horas_habiles(desde: datetime, horas: float, calendario: Calendario | None = None) -> datetime:
    """Avanza `horas` de reloj laboral desde un momento cualquiera."""
    cal = calendario or _CAL
    if horas < 0:
        raise ErrorDeValidacion("no se puede sumar un SLA negativo", campo="horas")
    original = desde.tzinfo or cal.huso
    momento = _siguiente_apertura(a_local(desde, cal), cal)
    restante = timedelta(hours=horas)

    while restante > timedelta(0):
        fin_del_dia = momento.replace(
            hour=cal.cierre.hour, minute=cal.cierre.minute, second=0, microsecond=0
        )
        disponible = fin_del_dia - momento
        if restante <= disponible:
            return (momento + restante).astimezone(original)
        restante -= disponible
        momento = _siguiente_apertura(fin_del_dia, cal)
    return momento.astimezone(original)


def sumar_dias_habiles(desde: datetime, dias: int, calendario: Calendario | None = None) -> datetime:
    cal = calendario or _CAL
    return sumar_horas_habiles(desde, dias * cal.horas_por_dia, cal)


def vencimiento(desde: datetime, criticidad: str, calendario: Calendario | None = None) -> datetime:
    cal = calendario or _CAL
    reglas = cargar_sla()
    if criticidad not in reglas:
        raise ErrorDeValidacion(f"criticidad desconocida: {criticidad!r}", campo="criticidad")
    return sumar_horas_habiles(desde, reglas[criticidad].horas(cal), cal)


@dataclass(frozen=True)
class Vencimiento:
    trace_id: str
    criticidad: str
    vence_en: str
    accion: Literal["escalar", "expirar", "bloquear"]
    escalamientos_previos: int
    motivo: str

    def as_dict(self) -> dict[str, object]:
        return {
            "trace_id": self.trace_id,
            "criticidad": self.criticidad,
            "vence_en": self.vence_en,
            "accion": self.accion,
            "escalamientos_previos": self.escalamientos_previos,
            "motivo": self.motivo,
        }


def resolver_vencimiento(
    *,
    trace_id: str,
    criticidad: str,
    espera_desde: datetime,
    ahora: datetime,
    escalamientos: int = 0,
    calendario: Calendario | None = None,
) -> Vencimiento | None:
    """Que hacer con un HITL que espera. `None` si todavia no vence.

    La salida nunca es "aprobar" — por eso la funcion no tiene ese caso ni puede tenerlo.
    Con la politica por defecto: alta escala y deja el caso bloqueado; media escala una vez
    y luego expira; baja expira.
    """
    cal = calendario or _CAL
    reglas = cargar_sla()
    if criticidad not in reglas:
        raise ErrorDeValidacion(f"criticidad desconocida: {criticidad!r}", campo="criticidad")

    limite = vencimiento(espera_desde, criticidad, cal)
    if ahora < limite:
        return None

    regla = reglas[criticidad]
    accion = regla.al_vencer if escalamientos == 0 else regla.luego
    motivos = {
        "escalar": "vencio el SLA: escala al siguiente nivel",
        "expirar": "vencio y ya se habia escalado: expira y se cierra como no_atendido"
        if escalamientos
        else "vencio el SLA de criticidad baja: expira y se cierra como no_atendido",
        "bloquear": "vencio despues de escalar: el caso queda bloqueado hasta que una persona lo resuelva",
    }

    if accion not in ACCIONES:
        raise NuncaAutoAprueba(f"accion invalida al vencer: {accion!r}")

    return Vencimiento(
        trace_id=trace_id,
        criticidad=criticidad,
        vence_en=limite.isoformat(timespec="seconds"),
        accion=accion,
        escalamientos_previos=escalamientos,
        motivo=motivos[accion],
    )

File: web/lib/reglas/sla.ts
/**
 * Puerto de `services/runlog/sla.py` — SLA y regla dura de timeout (arquitectura v3 §7.3).
 *
 *     Un HITL vencido escala o expira. Nunca auto-aprueba.
 *
 * ESTE ARCHIVO NO DECLARA NINGUN NUMERO. La jornada, los dias habiles, los festivos y los
 * plazos por criticidad viven en YAML y los lee `lib/registro/politicas.ts`:
 *
 *     registry/policies/calendario-laboral.yaml
 *     registry/policies/authority-gate.yaml -> hitl.sla
 *
 * Aqui solo vive la aritmetica, y existe porque Vercel no ejecuta Python. Dos
 * implementaciones de la misma regla divergen en silencio salvo que algo las ate: lo que
 * las ata es `tests/fixtures/sla-vectores.json`, generado desde el Python real por
 * `scripts/vectores_sla.py` y verificado en cada push por `.github/workflows/validar.yml`.
 * El fixture incluye ademas la configuracion con la que se genero, asi que el test no solo
 * comprueba que los resultados coinciden: comprueba que ambos leen los mismos YAML.
 *
 * SOBRE EL HUSO. El registro guarda `ts` en UTC y Vercel ejecuta en UTC, pero la jornada
 * es hora local de la empresa. Por eso aqui NO se usa `getHours()` ni ningun getter local
 * de `Date` — dependen del huso de la maquina, que en tu portatil y en Vercel no son el
 * mismo. Se trabaja en milisegundos y se desplaza el instante a "hora de pared local"
 * sumando el offset, leyendo despues con getters UTC, deterministas en todas partes.
 */

import {
  cargarCalendario,
  cargarSla,
  NuncaAutoAprueba,
  type Accion,
  type Calendario,
  type ReglaSLA,
} from "../registro/politicas";

export { NuncaAutoAprueba };
export type { Accion, Calendario, ReglaSLA };

const MS_DIA = 24 * 60 * 60 * 1000;
const MS_HORA = 60 * 60 * 1000;

export class CriticidadDesconocida extends Error {}

// La politica se lee una vez por proceso. En Vercel cada invocacion arranca en frio, asi
// que un cambio de YAML entra con el siguiente despliegue — que es exactamente cuando debe
// entrar: la politica cambia por PR, no en caliente.
let calendarioPorDefecto: Calendario | null = null;
let slaPorDefecto: Record<string, ReglaSLA> | null = null;

export const calendario = (): Calendario =>
  (calendarioPorDefecto ??= cargarCalendario());

export const reglas = (): Record<string, ReglaSLA> => (slaPorDefecto ??= cargarSla());

/** Solo para pruebas: obliga a releer los YAML en la siguiente llamada. */
export const olvidarPoliticas = (): void => {
  calendarioPorDefecto = null;
  slaPorDefecto = null;
};

// --- aritmetica del calendario ---------------------------------------------
// Estas funciones trabajan en "milisegundos locales": el instante UTC ya desplazado, de
// modo que leerlo con getters UTC devuelve la hora de pared de oficina.

const aLocal = (utcMs: number, cal: Calendario): number => utcMs + cal.offsetMs;
const aUtc = (localMs: number, cal: Calendario): number => localMs - cal.offsetMs;

/** Milisegundos transcurridos del dia. El doble modulo cubre fechas anteriores a 1970. */
const msDelDia = (localMs: number): number => ((localMs % MS_DIA) + MS_DIA) % MS_DIA;

const inicioDelDia = (localMs: number): number => localMs - msDelDia(localMs);

/** "AAAA-MM-DD" de la fecha local, para comparar contra la lista de festivos. */
const fechaLocal = (localMs: number): string =>
  new Date(localMs).toISOString().slice(0, 10);

/**
 * Dia habil declarado y que no sea festivo. Un festivo cuenta igual que un sabado.
 *
 * `getUTCDay()` cuenta 0=domingo; el YAML y Python cuentan 0=lunes. La conversion esta
 * aqui, en un solo lugar, y no repartida por el archivo.
 */
const esDiaHabil = (localMs: number, cal: Calendario): boolean => {
  const diaPython = (new Date(localMs).getUTCDay() + 6) % 7;
  return cal.diasHabiles.has(diaPython) && !cal.festivos.has(fechaLocal(localMs));
};

export const esHabil = (utcMs: number, cal: Calendario = calendario()): boolean => {
  const local = aLocal(utcMs, cal);
  const t = msDelDia(local);
  return esDiaHabil(local, cal) && t >= cal.aperturaMs && t < cal.cierreMs;
};

/** El siguiente instante en que la oficina esta abierta, o el mismo si ya lo esta. */
const siguienteApertura = (localMs: number, cal: Calendario): number => {
  let c = localMs;
  const t = msDelDia(c);
  if (t >= cal.cierreMs) {
    c = inicioDelDia(c) + MS_DIA + cal.aperturaMs;
  } else if (t < cal.aperturaMs) {
    c = inicioDelDia(c) + cal.aperturaMs;
  }
  // Un instante dentro de la jornada pero en sabado —o en festivo— no entra en ninguna
  // rama de arriba: lo mueve este bucle, igual que en el Python.
  while (!esDiaHabil(c, cal)) {
    c = inicioDelDia(c) + MS_DIA + cal.aperturaMs;
  }
  return c;
};

/** Avanza `horas` de reloj laboral desde un momento cualquiera. Entra y sale en UTC. */
export const sumarHorasHabiles = (
  utcMs: number,
  horas: number,
  cal: Calendario = calendario(),
): number => {
  if (horas < 0) throw new RangeError("no se puede sumar un SLA negativo");
  let momento = siguienteApertura(aLocal(utcMs, cal), cal);
  let restante = horas * MS_HORA;

  while (restante > 0) {
    const finDelDia = inicioDelDia(momento) + cal.cierreMs;
    const disponible = finDelDia - momento;
    if (restante <= disponible) return aUtc(momento + restante, cal);
    restante -= disponible;
    momento = siguienteApertura(finDelDia, cal);
  }
  return aUtc(momento, cal);
};

export const sumarDiasHabiles = (
  utcMs: number,
  dias: number,
  cal: Calendario = calendario(),
): number => sumarHorasHabiles(utcMs, dias * cal.horasPorDia, cal);

/** Cuantas horas de reloj laboral concede la politica a esta criticidad. */
const horasDe = (regla: ReglaSLA, cal: Calendario): number =>
  regla.horasHabiles ?? (regla.diasHabiles ?? 0) * cal.horasPorDia;

export const vencimiento = (
  utcMs: number,
  criticidad: string,
  cal: Calendario = calendario(),
  sla: Record<string, ReglaSLA> = reglas(),
): number => {
  const regla = sla[criticidad];
  if (!regla) throw new CriticidadDesconocida(`criticidad desconocida: ${criticidad}`);
  return sumarHorasHabiles(utcMs, horasDe(regla, cal), cal);
};

// --- la decision al vencer --------------------------------------------------

export interface Vencimiento {
  traceId: string;
  criticidad: string;
  venceEn: number;
  accion: Accion;
  escalamientosPrevios: number;
  motivo: string;
}

const MOTIVOS = {
  escalar: "vencio el SLA: escala al siguiente nivel",
  expirarPrimera: "vencio el SLA de criticidad baja: expira y se cierra como no_atendido",
  expirarTrasEscalar: "vencio y ya se habia escalado: expira y se cierra como no_atendido",
  bloquear:
    "vencio despues de escalar: el caso queda bloqueado hasta que una persona lo resuelva",
} as const;

/**
 * Que hacer con un HITL que espera. `null` si todavia no vence.
 *
 * La salida nunca es "aprobar" — el tipo `Accion` no la contiene y `cargarSla()` rechaza
 * una politica que la declare. Con la politica vigente: alta escala y deja el caso
 * bloqueado; media escala una vez y luego expira; baja expira.
 */
export const resolverVencimiento = (args: {
  traceId: string;
  criticidad: string;
  esperaDesde: number;
  ahora: number;
  escalamientos?: number;
  calendario?: Calendario;
  sla?: Record<string, ReglaSLA>;
}): Vencimiento | null => {
  const cal = args.calendario ?? calendario();
  const sla = args.sla ?? reglas();
  const escalamientos = args.escalamientos ?? 0;

  const regla = sla[args.criticidad];
  if (!regla) throw new CriticidadDesconocida(`criticidad desconocida: ${args.criticidad}`);

  const limite = vencimiento(args.esperaDesde, args.criticidad, cal, sla);
  if (args.ahora < limite) return null;

  const accion: Accion = escalamientos === 0 ? regla.alVencer : regla.luego;

  let motivo: string;
  if (accion === "escalar") motivo = MOTIVOS.escalar;
  else if (accion === "bloquear") motivo = MOTIVOS.bloquear;
  else motivo = escalamientos ? MOTIVOS.expirarTrasEscalar : MOTIVOS.expirarPrimera;

  return {
    traceId: args.traceId,
    criticidad: args.criticidad,
    venceEn: limite,
    accion,
    escalamientosPrevios: escalamientos,
    motivo,
  };
};

/**
 * El calendario del YAML con los festivos declarados en Postgres anadidos.
 *
 * Existe porque los festivos dejaron de vivir en `calendario-laboral.yaml` (ver
 * `scripts/sql/0003`): el huso, la jornada y los dias habiles siguen siendo politica en git,
 * pero la lista de feriados es un catalogo de la empresa que se captura desde el portal.
 *
 * Se pasa por parametro y no se lee aqui a proposito: leerlos es asincrono —una consulta— y
 * estas funciones son sincronas. Quien pinta la pagina los trae; quien los trae tambien se
 * asegura de que **no poder leerlos** sea un error visible y no una lista vacia, porque una
 * lista vacia es un calendario que afirma que se trabaja todos los dias y acorta el SLA sin
 * decirlo.
 */
export const conFestivos = (
  base: Calendario,
  declarados: Iterable<string>,
): Calendario => {
  // Union, no reemplazo: lo que quede escrito en el YAML sigue contando. Que la fuente se
  // haya movido no invalida lo que alguien ya habia declarado ahi.
  const fechas = new Set(base.festivos);
  for (const f of declarados) fechas.add(String(f).slice(0, 10));
  return { ...base, festivos: fechas };
};
